mixer/pkg/protobuf/yaml/dynamic/auth.go

```diff
@@ -0,0 +1,225 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynamic
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/oauth"
+
+	policypb "istio.io/api/policy/v1beta1"
+	"istio.io/istio/pkg/log"
+)
+
+const (
+	//
+	publicCACertsFile = "/etc/ssl/certs/ca-certificates.crt"
+)
+
+type authHelper struct {
+	authCfg          *policypb.Authentication
+	skipVerification bool
+}
+
+type headerToken struct {
+	header string
+	prefix string
+	token  []byte
+}
+
+func (c *headerToken) GetRequestMetadata(context.Context, ...string) (map[string]string, error) {
+	if c.prefix == "" {
+		return map[string]string{
+			c.header: string(c.token),
+		}, nil
+	}
+	return map[string]string{
+		c.header: c.prefix + " " + string(c.token),
+	}, nil
+}
+
+func (c *headerToken) RequireTransportSecurity() bool {
+	return true
+}
+
+func loadCerts(caFile string) (*x509.CertPool, error) {
+	// load ca cert.
+	caCertPool := x509.NewCertPool()
+	// load public root cert.
+	publicCerts, err := ioutil.ReadFile(publicCACertsFile)
+	if err != nil {
+		log.Warnf("Cannot load public root cert: %v", err)
+	} else {
+		caCertPool.AppendCertsFromPEM(publicCerts)
+	}
+
+	// load ca file provided by file.
+	if caFile != "" {
+		caCerts, err := ioutil.ReadFile(caFile)
+		if err != nil {
+			return nil, err
+		}
+		caCertPool.AppendCertsFromPEM(caCerts)
+	}
+	return caCertPool, nil
+}
+
+func buildMTLSDialOption(mtlsCfg *policypb.Mutual, skipVerify bool) ([]grpc.DialOption, error) {
+	// load peer cert/key.
+	pk := mtlsCfg.GetPrivateKey()
+	cc := mtlsCfg.GetClientCertificate()
+	peerCert, err := tls.LoadX509KeyPair(cc, pk)
+	if err != nil {
+		return nil, errors.Errorf("load peer cert/key error: %v", err)
+	}
+
+	// Load certificates, include public certs and certs from config.
+	caCertPool, err := loadCerts(mtlsCfg.GetCaCertificates())
+	if err != nil {
+		return nil, fmt.Errorf("failed build tls dial option: ca cert cannot be load %v", err)
+	}
+
+	tc := credentials.NewTLS(&tls.Config{
+		Certificates:       []tls.Certificate{peerCert},
+		RootCAs:            caCertPool,
+		InsecureSkipVerify: skipVerify,
+	})
+	return []grpc.DialOption{grpc.WithTransportCredentials(tc)}, nil
+}
+
+func buildOAuthToken(oauthCfg *policypb.OAuth) (oauth2.TokenSource, error) {
+	sp := oauthCfg.ClientSecret
+	if sp == "" {
+		return nil, errors.New("oauth secret cannot be empty")
+	}
+	secret, err := ioutil.ReadFile(sp)
+	if err != nil {
+		return nil, err
+	}
+	c := &clientcredentials.Config{
+		ClientID:     oauthCfg.ClientId,
+		ClientSecret: string(secret),
+		TokenURL:     oauthCfg.TokenUrl,
+		Scopes:       oauthCfg.Scopes,
+	}
+	return c.TokenSource(context.Background()), nil
```

Comment:
nit: context would be useful to pass in, since it allows interrupting hanging HTTP connections.

Reply:
I am struggled to find the proper context to pass in here: https://github.com/istio/istio/blob/master/mixer/pkg/protobuf/yaml/dynamic/handler.go#L149. Any suggestion?

Reply:
There is https://godoc.org/google.golang.org/grpc#DialContext, but I think it should be handled in a separate PR. Just leave a TODO here to clean it up and wire context from gRPC.